lib/mix/lib/mix/tasks/xref.ex

```diff
@@ -283,6 +320,85 @@ defmodule Mix.Tasks.Xref do
     Mix.raise message
   end
 
+  ## Graph helpers
+
+  defp excluded(opts) do
+    Keyword.get_values(opts, :exclude)
+    |> Enum.map(&{&1, nil})
+  end
+
+  defp module_references() do
+    source_modules =
+      for manifest <- E.manifests(),
+          manifest_data = read_manifest(manifest, ""),
+          module(module: module, source: source) <- manifest_data,
+          source = Enum.find(manifest_data, &match?(source(source: ^source), &1)),
+          do: {source, module}
+
+    source_modules = Enum.group_by(source_modules, &elem(&1, 0), &elem(&1, 1))
+
+    all_modules =
+      Enum.reduce source_modules, MapSet.new(), fn {_, modules}, acc ->
+        Enum.reduce(modules, acc, &MapSet.put(&2, &1))
+      end
+
+    Enum.reduce source_modules, [], fn {source, modules}, acc ->
+      source(runtime_references: r, compile_references: c) = source
```

Comment:
Yes, we certainly could go by file instead. It's really just noise to list more than one entry per file.